File: ios-native/KPKNFit/KPKNFit/Domain/Templates/SessionTemplateCatalogPolicy.swift
import Foundation

struct SplitTemplateDayGroup {
    let splitId: String
    let splitName: String
    let dayIndex: Int
    let dayLabel: String
    let templates: [SessionTemplate]
    let estimatedMuscleVolume: [String: Double]
    let warnings: [String]
}

struct FocusTemplateGroup {
    let category: SessionTemplateFocusCategory
    let templates: [SessionTemplate]
}

struct UserTemplateGroup {
    let templates: [SessionTemplate]
}

enum SessionTemplateCatalogPolicy {

    static let weeklyVolumeRanges: [String: ClosedRange<Double>] = [
        "Pectorales": 10.0...14.0,
        "Dorsales": 10.0...14.0,
        "Trapecio": 6.0...10.0,
        "Deltoides": 8.0...12.0,
        "Bíceps": 6.0...10.0,
        "Tríceps": 6.0...10.0,
        "Cuádriceps": 10.0...14.0,
        "Isquiosurales": 8.0...12.0,
        "Glúteos": 8.0...14.0,
        "Pantorrillas": 6.0...12.0,
        "Aductores": 4.0...8.0,
        "Abdomen": 4.0...10.0,
        "Core": 4.0...8.0,
        "Erectores Espinales": 3.0...6.0,
    ]

    static func templatesForSplit(
        _ split: SplitTemplate,
        templates: [SessionTemplate],
        exerciseIndex: [String: ExerciseMuscleInfo] = [:]
    ) -> [SplitTemplateDayGroup] {
        guard split.id != "custom" else { return [] }
        var groups: [SplitTemplateDayGroup] = []

        for (index, dayLabel) in split.pattern.enumerated() {
            if dayLabel.caseInsensitiveCompare("Descanso") == .orderedSame { continue }
            let dayTemplates = templatesForSplitDay(splitId: split.id, dayLabel: dayLabel, templates: templates)
            let representative = dayTemplates.first
            let estimatedVol: [String: Double]
            if let rep = representative, !exerciseIndex.isEmpty {
                estimatedVol = calculateSessionMuscleVolume(session: rep.session, exerciseIndex: exerciseIndex)
            } else {
                estimatedVol = [:]
            }

            var warnings: [String] = []
            if let rep = representative, !exerciseIndex.isEmpty {
                let drain = evaluateTemplateRings(template: rep, exerciseIndex: exerciseIndex)
                let isPl = isPowerliftingTemplate(rep)
                let maxCns = isPl ? 45 : 35
                let maxMuscular = isPl ? 50 : 45
                let maxSpinal = isPl ? 40 : 30
                if drain.cns > maxCns { warnings.append("Fatiga SNC elevada (\(drain.cns)% > \(maxCns)%)") }
                if drain.muscular > maxMuscular { warnings.append("Fatiga Muscular elevada (\(drain.muscular)% > \(maxMuscular)%)") }
                if drain.spinal > maxSpinal { warnings.append("Carga axial/espinal elevada (\(drain.spinal)% > \(maxSpinal)%)") }
            }

            groups.append(SplitTemplateDayGroup(
                splitId: split.id, splitName: split.name,
                dayIndex: index, dayLabel: dayLabel,
                templates: dayTemplates.sorted { $0.sortOrder < $1.sortOrder },
                estimatedMuscleVolume: estimatedVol, warnings: warnings
            ))
        }
        return groups
    }

    static func independentTemplateGroups(_ templates: [SessionTemplate]) -> [FocusTemplateGroup] {
        let independent = templates.filter { $0.splitIds.isEmpty && $0.focusCategory != nil }
        return Dictionary(grouping: independent) { $0.focusCategory! }
            .map { FocusTemplateGroup(category: $0.key, templates: $0.value.sorted { $0.sortOrder < $1.sortOrder }) }
            .sorted { $0.category.rawValue < $1.category.rawValue }
    }

    static func userTemplateGroup(_ templates: [SessionTemplate]) -> UserTemplateGroup {
        UserTemplateGroup(templates: templates
            .filter { $0.sourceType == .user && !$0.isArchived }
            .sorted { $0.name.lowercased() < $1.name.lowercased() })
    }

    static func templatesForSplitDay(splitId: String, dayLabel: String, templates: [SessionTemplate]) -> [SessionTemplate] {
        let normalizedCandidates = candidateLabelsFor(dayLabel)
        let exact = templates.filter { $0.splitIds.contains(splitId) && $0.splitDayLabels.contains { $0.caseInsensitiveCompare(dayLabel) == .orderedSame } }
        let sameSplitArchetype = templates.filter { $0.splitIds.contains(splitId) && $0.splitDayLabels.contains { normalizedCandidates.contains($0.normalizedLabel()) } }
        let sharedArchetype = templates.filter { !$0.splitIds.isEmpty && $0.splitDayLabels.contains { normalizedCandidates.contains($0.normalizedLabel()) } }
        let independentArchetype = templates.filter { $0.splitIds.isEmpty && $0.focusCategory != nil && focusCategoriesFor(dayLabel).contains($0.focusCategory!) }

        return (exact.map { ($0, 1) } + sameSplitArchetype.map { ($0, 2) } + sharedArchetype.map { ($0, 3) } + independentArchetype.map { ($0, 4) })
            .uniqued(by: { $0.0.id })
            .sorted { $0.1 < $1.1 || ($0.1 == $1.1 && $0.0.sortOrder < $1.0.sortOrder) }
            .map { $0.0 }
    }

    static func calculateSessionMuscleVolume(session: Session, exerciseIndex: [String: ExerciseMuscleInfo]) -> [String: Double] {
        let entries = VolumeCalculator.calculateUnifiedMuscleVolume(sessions: [session], allExerciseInfo: Array(exerciseIndex.values))
        return Dictionary(uniqueKeysWithValues: entries.map { ($0.muscleName, $0.displayVolume) })
    }

    static func evaluateTemplateRings(template: SessionTemplate, exerciseIndex: [String: ExerciseMuscleInfo]) -> PredictedDrain {
        (try? AugeFatigueEngine.calculateAdjustedPredictedDrain(session: template.session, exerciseIndex: exerciseIndex, settings: Settings())) ?? PredictedDrain(cns: 100, muscular: 100, spinal: 100)
    }

    static func isPowerliftingTemplate(_ template: SessionTemplate) -> Bool {
        template.focusCategory == .powerlifting ||
            template.tags.contains { $0.name.contains("POWERLIFTING") || $0.name.contains("SENTADILLA") || $0.name.contains("PESO_MUERTO") || $0.name.contains("BANCA") } ||
            template.id.contains("pl-") || template.id.contains("sbd") || template.id.contains("sheiko") || template.id.contains("texas")
    }

    private static func candidateLabelsFor(_ dayLabel: String) -> Set<String> {
        let normalized = dayLabel.normalizedLabel()
        var labels: [String] = [normalized]

        func add(_ values: String...) { values.forEach { labels.append($0.normalizedLabel()) } }

        if normalized.contains("empuje") || normalized.contains("push") || normalized.contains("press") || (normalized.contains("pecho") && normalized.contains("triceps")) {
            add("Empuje", "Pecho", "Push", "Press Banca/Pecho", "Pecho/Espalda")
        }
        if normalized.contains("tirón") || normalized.contains("tiron") || normalized.contains("pull") || normalized.contains("espalda") || normalized.contains("bíceps") || normalized.contains("biceps") {
            add("Tirón", "Espalda", "Pull", "Peso Muerto/Espalda", "Espalda/Bíceps", "Cadena Posterior")
        }
        if normalized.contains("pierna") || normalized.contains("lower") || normalized.contains("cuádriceps") || normalized.contains("cuadriceps") {
            add("Pierna", "Piernas", "Lower", "Cuádriceps/Glúteo", "Pierna Mantenimiento", "Sentadilla/Pierna", "Cadena Anterior")
        }
        if normalized.contains("gluteo") || normalized.contains("glúteo") {
            add("Glúteo/Isquios", "Glúteo Pump", "Cuádriceps/Glúteo", "Pierna", "Lower")
        }
        if normalized.contains("torso") || normalized.contains("upper") {
            add("Torso", "Upper", "Upper Completo", "Torso Liviano")
        }
        if normalized.contains("full body") || normalized.contains("cuerpo completo") {
            add("Cuerpo Completo", "Full Body", "Full Body A", "Full Body B", "Cuerpo Completo A", "Cuerpo Completo B", "Cuerpo Completo C", "Cuerpo Completo D", "Full Body Pesado", "Full Body Liviano", "Full Body Medio")
        }
        if normalized.contains("hombro") || normalized.contains("brazo") || normalized.contains("triceps") || normalized.contains("biceps") {
            add("Hombro/Brazo", "Hombros/Brazos", "Brazos/Hombros", "Hombros/Abs", "T1 Militar")
        }
        if normalized.contains("anterior") { add("Cadena Anterior") }
        if normalized.contains("posterior") { add("Cadena Posterior") }
        if normalized.contains("sbd") { add("SBD Día 1", "SBD Día 2", "SBD Día 3", "SBD (Volumen)", "SBD (Técnica)", "SBD (Intensidad)") }
        if normalized.contains("sentadilla") || normalized.contains("squat") { add("Sentadilla/Banca", "Sentadilla/Peso Muerto", "T1 Sentadilla", "Sentadilla") }
        if normalized.contains("banca") || normalized.contains("bench") { add("Banca Volumen", "Banca", "T1 Banca", "Banca/Sentadilla Var.", "Peso Muerto/Banca Var.") }
        if normalized.contains("peso muerto") || normalized.contains("deadlift") || normalized.contains("dl") { add("Peso Muerto", "Peso Muerto/Banca", "Peso Muerto/Accesorios", "Peso Muerto/Press", "Variante DL/Banca", "T1 Peso Muerto") }
        if normalized.contains("volumen") || normalized.contains("recuperación") || normalized.contains("recuperacion") || normalized.contains("intensidad") || normalized.contains("repeticiones") || normalized.contains("explosivo") || normalized.contains("pesado") || normalized.contains("liviano") || normalized.contains("moderado") {
            add("Full Body", "SBD Día 1", "SBD Día 2", "SBD Día 3", "Minimalista · Fuerza")
        }
        if normalized.contains("max effort") || normalized.contains("dynamic effort") || normalized.contains("me ") || normalized.contains("de ") { add("Sentadilla/Banca", "T1 Sentadilla", "T1 Banca") }
        if normalized.contains("bodybuilding") || normalized.contains("accesorios") { add("Upper", "Torso", "Hombro/Brazo") }
        if normalized.contains("sesion") { add("SBD Día 1", "SBD Día 2", "SBD Día 3", "Full Body", "Minimalista · Fuerza") }

        return Set(labels)
    }

    private static func focusCategoriesFor(_ dayLabel: String) -> Set<SessionTemplateFocusCategory> {
        let normalized = dayLabel.normalizedLabel()
        var categories = Set<SessionTemplateFocusCategory>()
        if normalized.contains("pierna") || normalized.contains("lower") || normalized.contains("cuádriceps") || normalized.contains("cuadriceps") { categories.insert(.piernas) }
        if normalized.contains("gluteo") || normalized.contains("glúteo") { categories.insert(.gluteos) }
        if normalized.contains("pecho") || normalized.contains("empuje") || normalized.contains("push") || normalized.contains("banca") { categories.insert(.pecho) }
        if normalized.contains("espalda") || normalized.contains("tirón") || normalized.contains("tiron") || normalized.contains("pull") { categories.insert(.espalda) }
        if normalized.contains("hombro") { categories.insert(.hombros) }
        if normalized.contains("brazo") || normalized.contains("biceps") || normalized.contains("bíceps") || normalized.contains("triceps") { categories.insert(.brazos) }
        if normalized.contains("full") || normalized.contains("cuerpo completo") || normalized.contains("volumen") || normalized.contains("recuperación") || normalized.contains("recuperacion") { categories.insert(.fullBody) }
        if normalized.contains("sbd") || normalized.contains("sentadilla") || normalized.contains("banca") || normalized.contains("peso muerto") || normalized.contains("t1") || normalized.contains("max effort") || normalized.contains("dynamic effort") { categories.insert(.powerlifting) }
        if normalized.contains("minimalista") { categories.insert(.minimalista) }
        if normalized.contains("anterior") || normalized.contains("posterior") || normalized.contains("sesion") { categories.insert(.fullBody) }
        if normalized.contains("recuperación") || normalized.contains("recuperacion") || normalized.contains("light") { categories.insert(.recuperacion) }
        return categories
    }
}

private extension String {
    func normalizedLabel() -> String {
        self.lowercased()
            .replacingOccurrences(of: "á", with: "a")
            .replacingOccurrences(of: "é", with: "e")
            .replacingOccurrences(of: "í", with: "i")
            .replacingOccurrences(of: "ó", with: "o")
            .replacingOccurrences(of: "ú", with: "u")
            .trimmingCharacters(in: .whitespaces)
    }
}

File: android-native/app/src/main/java/com/example/kpkn/domain/templates/SessionTemplateCatalogPolicy.kt
package com.example.kpkn.domain.templates

import com.example.kpkn.data.models.Exercise
import com.example.kpkn.data.models.ExerciseMuscleInfo
import com.example.kpkn.data.models.PredictedDrain
import com.example.kpkn.data.models.Session
import com.example.kpkn.data.models.Settings
import com.example.kpkn.data.sessions.SessionTemplate
import com.example.kpkn.data.sessions.SessionTemplateFocusCategory
import com.example.kpkn.data.sessions.SessionTemplateSourceType
import com.example.kpkn.data.splits.SplitTemplate
import com.example.kpkn.domain.auge.AugeFatigueEngine
import com.example.kpkn.domain.training.VolumeCalculator

data class SplitTemplateDayGroup(
    val splitId: String,
    val splitName: String,
    val dayIndex: Int,
    val dayLabel: String,
    val templates: List<SessionTemplate>,
    val estimatedMuscleVolume: Map<String, Double> = emptyMap(),
    val warnings: List<String> = emptyList()
)

data class FocusTemplateGroup(
    val category: SessionTemplateFocusCategory,
    val templates: List<SessionTemplate>
)

data class UserTemplateGroup(
    val templates: List<SessionTemplate>
)

object SessionTemplateCatalogPolicy {

    // Rango de volumen semanal objetivo por músculo canónico
    val WEEKLY_VOLUME_RANGES = mapOf(
        "Pectorales" to 10.0..14.0,
        "Dorsales" to 10.0..14.0,
        "Trapecio" to 6.0..10.0,
        "Deltoides" to 8.0..14.0,
        "Bíceps" to 6.0..10.0,
        "Tríceps" to 6.0..10.0,
        "Cuádriceps" to 10.0..16.0,
        "Isquiosurales" to 8.0..12.0,
        "Glúteos" to 8.0..14.0,
        "Pantorrillas" to 6.0..12.0,
        "Aductores" to 4.0..8.0,
        "Abdomen" to 4.0..10.0,
        "Core" to 4.0..8.0,
        // Compuestos (sentadilla / bisagra) aportan erectores secundarios reales del catálogo.
        "Erectores Espinales" to 3.0..12.0,
    )

    fun templatesForSplit(
        split: SplitTemplate,
        templates: List<SessionTemplate>,
        exerciseIndex: Map<String, ExerciseMuscleInfo> = emptyMap()
    ): List<SplitTemplateDayGroup> {
        if (split.id == "custom") return emptyList()
        val groups = mutableListOf<SplitTemplateDayGroup>()

        split.pattern.forEachIndexed { index, dayLabel ->
            if (dayLabel.equals("Descanso", ignoreCase = true)) return@forEachIndexed

            val dayTemplates = templatesForSplitDay(split.id, dayLabel, templates)

            // We pre-calculate estimated muscle volumes and warnings for the first template as representative
            val representativeTemplate = dayTemplates.firstOrNull()
            val estimatedVol = if (representativeTemplate != null && exerciseIndex.isNotEmpty()) {
                calculateSessionMuscleVolume(representativeTemplate.session, exerciseIndex)
            } else {
                emptyMap()
            }

            val warnings = mutableListOf<String>()
            if (representativeTemplate != null && exerciseIndex.isNotEmpty()) {
                val drain = evaluateTemplateRings(representativeTemplate, exerciseIndex)
                val isPl = isPowerliftingTemplate(representativeTemplate)
                val caps = RingBudgetPolicy.sessionWarningCaps(isPl)

                if (drain.cns > caps.cns) warnings += "Fatiga SNC elevada (${drain.cns}% > ${caps.cns}%)"
                if (drain.muscular > caps.muscular) warnings += "Fatiga Muscular elevada (${drain.muscular}% > ${caps.muscular}%)"
                if (drain.spinal > caps.spinal) warnings += "Carga axial/espinal elevada (${drain.spinal}% > ${caps.spinal}%)"
            }

            groups.add(
                SplitTemplateDayGroup(
                    splitId = split.id,
                    splitName = split.name,
                    dayIndex = index,
                    dayLabel = dayLabel,
                    templates = dayTemplates.sortedBy { it.sortOrder },
                    estimatedMuscleVolume = estimatedVol,
                    warnings = warnings
                )
            )
        }
        return groups
    }

    fun independentTemplateGroups(templates: List<SessionTemplate>): List<FocusTemplateGroup> {
        val independentTemplates = templates.filter { it.splitIds.isEmpty() && it.focusCategory != null }
        return independentTemplates
            .groupBy { it.focusCategory!! }
            .map { (category, list) ->
                FocusTemplateGroup(category, list.sortedBy { it.sortOrder })
            }
            .sortedBy { it.category.ordinal }
    }

    fun userTemplateGroup(templates: List<SessionTemplate>): UserTemplateGroup {
        return UserTemplateGroup(
            templates = templates
                .filter { it.sourceType == SessionTemplateSourceType.USER && !it.isArchived }
                .sortedBy { it.name.lowercase() }
        )
    }

    fun templatesForSplitDay(
        splitId: String,
        dayLabel: String,
        templates: List<SessionTemplate>
    ): List<SessionTemplate> {
        val normalizedCandidates = candidateLabelsFor(dayLabel)
        val exact = templates.filter { template ->
            template.splitIds.contains(splitId) && template.splitDayLabels.any { it.equals(dayLabel, ignoreCase = true) }
        }
        val sameSplitArchetype = templates.filter { template ->
            template.splitIds.contains(splitId) && template.splitDayLabels.any { it.normalizedLabel() in normalizedCandidates }
        }
        val sharedArchetype = templates.filter { template ->
            template.splitIds.isNotEmpty() && template.splitDayLabels.any { it.normalizedLabel() in normalizedCandidates }
        }
        val independentArchetype = templates.filter { template ->
            template.splitIds.isEmpty() && template.focusCategory != null && template.focusCategory in focusCategoriesFor(dayLabel)
        }

        val exactMapped = exact.map { it to 1 }
        val sameSplitMapped = sameSplitArchetype.map { it to 2 }
        val sharedMapped = sharedArchetype.map { it to 3 }
        val independentMapped = independentArchetype.map { it to 4 }

        return (exactMapped + sameSplitMapped + sharedMapped + independentMapped)
            .distinctBy { it.first.id }
            .sortedWith(compareBy<Pair<SessionTemplate, Int>> { it.second }.thenBy { it.first.sortOrder })
            .map { it.first }
    }

    fun calculateSessionMuscleVolume(
        session: Session,
        exerciseIndex: Map<String, ExerciseMuscleInfo>
    ): Map<String, Double> {
        val entries = VolumeCalculator.calculateUnifiedMuscleVolume(listOf(session), exerciseIndex.values.toList())
        return entries.associate { it.muscleName to it.displayVolume }
    }

    fun evaluateTemplateRings(
        template: SessionTemplate,
        exerciseIndex: Map<String, ExerciseMuscleInfo>,
        settings: Settings = Settings(),
    ): PredictedDrain {
        return try {
            AugeFatigueEngine.calculateAdjustedPredictedDrain(
                template.session, exerciseIndex, settings,
            )
        } catch (_: Throwable) {
            PredictedDrain(cns = 100, muscular = 100, spinal = 100)
        }
    }

    fun isPowerliftingTemplate(template: SessionTemplate): Boolean {
        return template.focusCategory == SessionTemplateFocusCategory.POWERLIFTING ||
                template.tags.any {
                    it.name.contains("POWERLIFTING") ||
                    it.name.contains("SENTADILLA") ||
                    it.name.contains("PESO_MUERTO") ||
                    it.name.contains("BANCA")
                } ||
                template.id.contains("pl-") ||
                template.id.contains("sbd") ||
                template.id.contains("sheiko") ||
                template.id.contains("texas")
    }

    fun candidateLabelsForDay(dayLabel: String): Set<String> = candidateLabelsFor(dayLabel)

    fun focusCategoriesForDay(dayLabel: String): Set<SessionTemplateFocusCategory> =
        focusCategoriesFor(dayLabel)

    private fun candidateLabelsFor(dayLabel: String): Set<String> {
        val normalized = dayLabel.normalizedLabel()
        val labels = linkedSetOf(normalized)
        fun add(vararg values: String) = values.forEach { labels += it.normalizedLabel() }

        if (normalized.contains("empuje") || normalized.contains("push") || normalized.contains("press") || normalized.contains("pecho") && normalized.contains("triceps")) {
            add("Empuje", "Pecho", "Push", "Press Banca/Pecho", "Pecho/Espalda")
        }
        if (normalized.contains("tirón") || normalized.contains("tiron") || normalized.contains("pull") || normalized.contains("espalda") || normalized.contains("bíceps") || normalized.contains("biceps")) {
            add("Tirón", "Espalda", "Pull", "Peso Muerto/Espalda", "Espalda/Bíceps", "Cadena Posterior")
        }
        if (normalized.contains("pierna") || normalized.contains("lower") || normalized.contains("cuádriceps") || normalized.contains("cuadriceps")) {
            add(
                "Pierna", "Piernas", "Lower", "Cuádriceps/Glúteo", "Pierna Mantenimiento",
                "Sentadilla/Pierna", "Cadena Anterior",
                "Isquios", "Isquiosurales", "Glúteos", "Glúteo", "Femoral", "Glúteo/Isquios",
            )
        }
        if (normalized.contains("isquio") || normalized.contains("femoral") || normalized.contains("hamstring")) {
            add("Isquios", "Isquiosurales", "Femoral", "Glúteo/Isquios", "Pierna", "Lower", "Cadena Posterior")
        }
        if (normalized.contains("gluteo") || normalized.contains("glúteo")) {
            add("Glúteo/Isquios", "Glúteo Pump", "Cuádriceps/Glúteo", "Pierna", "Lower", "Glúteos", "Glúteo")
        }
        if (normalized.contains("torso") || normalized.contains("upper")) {
            add(
                "Torso", "Upper", "Upper Completo", "Torso Liviano",
                "Pecho", "Espalda", "Hombros", "Brazos", "Pecho/Espalda", "Hombro/Brazo",
                "Hombros/Brazos", "Espalda/Bíceps", "Press Banca/Pecho",
            )
        }
        if (normalized.contains("full body") || normalized.contains("cuerpo completo")) {
            add("Cuerpo Completo", "Full Body", "Full Body A", "Full Body B", "Cuerpo Completo A", "Cuerpo Completo B", "Cuerpo Completo C", "Cuerpo Completo D", "Full Body Pesado", "Full Body Liviano", "Full Body Medio")
        }
        if (normalized.contains("hombro") || normalized.contains("delts") || normalized.contains("brazo") || normalized.contains("brazos") || normalized.contains("triceps") || normalized.contains("biceps")) {
            add("Hombro/Brazo", "Hombros/Brazos", "Brazos/Hombros", "Hombros/Abs", "T1 Militar", "Hombros", "Brazos")
        }
        if (normalized.contains("anterior")) add("Cadena Anterior")
        if (normalized.contains("posterior")) add("Cadena Posterior")

        if (normalized.contains("sbd")) add("SBD Día 1", "SBD Día 2", "SBD Día 3", "SBD (Volumen)", "SBD (Técnica)", "SBD (Intensidad)")
        if (normalized.contains("sentadilla") || normalized.contains("squat")) add("Sentadilla/Banca", "Sentadilla/Peso Muerto", "T1 Sentadilla", "Sentadilla")
        if (normalized.contains("banca") || normalized.contains("bench")) add("Banca Volumen", "Banca", "T1 Banca", "Banca/Sentadilla Var.", "Peso Muerto/Banca Var.")
        if (normalized.contains("peso muerto") || normalized.contains("deadlift") || normalized.contains("dl")) add("Peso Muerto", "Peso Muerto/Banca", "Peso Muerto/Accesorios", "Peso Muerto/Press", "Variante DL/Banca", "T1 Peso Muerto")
        if (normalized.contains("volumen") || normalized.contains("recuperación") || normalized.contains("recuperacion") || normalized.contains("intensidad") || normalized.contains("repeticiones") || normalized.contains("explosivo") || normalized.contains("pesado") || normalized.contains("liviano") || normalized.contains("moderado")) {
            add("Full Body", "SBD Día 1", "SBD Día 2", "SBD Día 3", "Minimalista · Fuerza")
        }
        if (normalized.contains("max effort") || normalized.contains("dynamic effort") || normalized.contains("me ") || normalized.contains("de ")) add("Sentadilla/Banca", "T1 Sentadilla", "T1 Banca")
        if (normalized.contains("bodybuilding") || normalized.contains("accesorios")) add("Upper", "Torso", "Hombro/Brazo")
        if (normalized.contains("sesion")) add("SBD Día 1", "SBD Día 2", "SBD Día 3", "Full Body", "Minimalista · Fuerza")

        return labels
    }

    private fun focusCategoriesFor(dayLabel: String): Set<SessionTemplateFocusCategory> {
        val normalized = dayLabel.normalizedLabel()
        val categories = linkedSetOf<SessionTemplateFocusCategory>()
        if (normalized.contains("pierna") || normalized.contains("lower") || normalized.contains("cuádriceps") || normalized.contains("cuadriceps")) {
            categories += SessionTemplateFocusCategory.PIERNAS
            categories += SessionTemplateFocusCategory.CUADRICEPS
            categories += SessionTemplateFocusCategory.ISQUIOS
            categories += SessionTemplateFocusCategory.GLUTEOS
        }
        if (normalized.contains("isquio") || normalized.contains("femoral") || normalized.contains("hamstring")) {
            categories += SessionTemplateFocusCategory.ISQUIOS
            categories += SessionTemplateFocusCategory.PIERNAS
        }
        if (normalized.contains("gluteo") || normalized.contains("glúteo")) categories += SessionTemplateFocusCategory.GLUTEOS
        if (normalized.contains("pecho") || normalized.contains("empuje") || normalized.contains("push") || normalized.contains("banca")) categories += SessionTemplateFocusCategory.PECHO
        if (normalized.contains("espalda") || normalized.contains("tirón") || normalized.contains("tiron") || normalized.contains("pull")) categories += SessionTemplateFocusCategory.ESPALDA
        if (normalized.contains("hombro") || normalized.contains("delts")) categories += SessionTemplateFocusCategory.HOMBROS
        if (normalized.contains("brazo") || normalized.contains("brazos") || normalized.contains("biceps") || normalized.contains("bíceps") || normalized.contains("triceps")) {
            categories += SessionTemplateFocusCategory.BRAZOS
        }
        if (normalized.contains("torso") || normalized.contains("upper")) {
            categories += SessionTemplateFocusCategory.PECHO
            categories += SessionTemplateFocusCategory.ESPALDA
            categories += SessionTemplateFocusCategory.HOMBROS
            categories += SessionTemplateFocusCategory.BRAZOS
        }
        if (normalized.contains("full") || normalized.contains("cuerpo completo") || normalized.contains("volumen") || normalized.contains("recuperación") || normalized.contains("recuperacion")) categories += SessionTemplateFocusCategory.FULL_BODY
        if (normalized.contains("sbd") || normalized.contains("sentadilla") || normalized.contains("banca") || normalized.contains("peso muerto") || normalized.contains("t1") || normalized.contains("max effort") || normalized.contains("dynamic effort")) categories += SessionTemplateFocusCategory.POWERLIFTING
        if (normalized.contains("minimalista")) categories += SessionTemplateFocusCategory.MINIMALISTA
        if (normalized.contains("anterior")) categories += SessionTemplateFocusCategory.CADENA_ANTERIOR
        if (normalized.contains("posterior")) categories += SessionTemplateFocusCategory.CADENA_POSTERIOR
        if (normalized.contains("pantorr") || normalized.contains("gemelo") || normalized.contains("calf")) {
            categories += SessionTemplateFocusCategory.PANTORRILLAS
        }
        if (normalized.contains("core") || normalized.contains("abdom") || normalized.contains("abs")) {
            categories += SessionTemplateFocusCategory.CORE
        }
        if (normalized.contains("antebrazo") || normalized.contains("forearm") || normalized.contains("grip")) {
            categories += SessionTemplateFocusCategory.ANTEBRAZOS
        }
        if (normalized.contains("aductor") || normalized.contains("adductor")) {
            categories += SessionTemplateFocusCategory.ADUCTORES
        }
        if (normalized.contains("sesion")) categories += SessionTemplateFocusCategory.FULL_BODY
        if (normalized.contains("recuperación") || normalized.contains("recuperacion") || normalized.contains("light")) categories += SessionTemplateFocusCategory.RECUPERACION
        return categories
    }

    private fun String.normalizedLabel(): String = lowercase()
        .replace("á", "a")
        .replace("é", "e")
        .replace("í", "i")
        .replace("ó", "o")
        .replace("ú", "u")
        .trim()
}
